src/compiler/crystal/semantic/warnings.cr

```diff
@@ -48,6 +49,30 @@ module Crystal
       end
     end
 
+    def check_deprecated_constant(const : Const, node : Path)
+      return unless self.warnings.all?
+
+      if (ann = const.annotation(self.deprecated_annotation)) &&
+         (deprecated_annotation = DeprecatedAnnotation.from(ann))
+        call_location = node.location.try(&.macro_location) || node.location
+
+        return if self.ignore_warning_due_to_location?(call_location)
+        warning_key = call_location.try { |l| "#{const} #{l}" }
+
+        # skip warning if the call site was already informed
+        # if there is no location information just inform it.
+        return if !warning_key || @deprecated_constants_detected.includes?(warning_key)
+        @deprecated_constants_detected.add(warning_key) if warning_key
```

Comment:
This early return condition appears to contradict the comment? The logic doesn't seem to make sense. But this is apparently copied from existing deprecation handlers, to I presume it's better to fix this outside this PR. I don't mind whether before or after.
In any case, I presume there could also be some more refactoring such as extracing common functionality from these methods.

Reply:
@HertzDevil Did you see this comment?